FindEx implementetation + refactoring, take 4

borisfom opened the pull request:
This is to replace https://github.com/soumith/cudnn.torch/pull/250. 
Restored functional.lua : @szagoruyko, please feel free to port it to find.lua using temporary 'layer' variables.
All tests pass.

borisfom commented:
@soumith: this PR is ready to go. functional.lua port to find is entirely optional.

soumith commented on RNN.lua:
```diff
@@ -330,18 +321,20 @@ function RNN:updateOutput(input)
-   workspaceSize[1] = torch.floor((workspaceSize[1] + 3) / 4) -- sizeof(float)
+   cudnn.setSharedWorkspaceSize(workspaceSize[1], true)
+   local wsPtr, wsSize = cudnn.getSharedWorkspace()
+--[[   local elemSize = self.workspace:elementSize()
```
please remove commented code.

borisfom replied:
Done now.

soumith commented on RNN.lua:
```diff
@@ -351,12 +344,13 @@ function RNN:updateOutput(input)
+      local elemSize = self.reserve:elementSize()
+      reserveSize[1] = torch.floor((reserveSize[1] + elemSize - 1) / elemSize) -- sizeof(float)
+      self.reserve:resize(reserveSize[1])
+--[[      if self.reserve:dim() == 0 or
```
remove commented code

szagoruyko commented on SpatialConvolution.lua:
```diff
@@ -361,24 +184,26 @@ local function makeContiguous(self, input, gradOutput)
+    local extraBuffer, extraBufferSize = cudnn.getSharedWorkspace()
     for g = 0, self.groups - 1 do
-        errcheck('cudnnConvolutionForward', cudnn.getHandle(),
+        errcheck(self,'cudnnConvolutionForward', cudnn.getHandle(),
```
having two different errcheck functions with the same name is confusing, can you give it a different name?

ngimel commented on find.lua:
```diff
@@ -64,93 +64,99 @@ local bwdDataAlgoNames = {
-   if find.verbose then
+local function getConvolutionDescriptor(desc)
+   local CUDNN_DIM_MAX=4
+
```
just do 
```
local dimMax='CUDNN_DIM_MAX'
```
here, it will pick CUDNN_DIM_MAX value from ffi.lua. Besides, CUDNN_DIM_MAX=4 is not enough for volumetric convolutions?

ngimel commented on find.lua:
```diff
@@ -64,93 +64,99 @@ local bwdDataAlgoNames = {
    return status
 end
-find.call = call
+find.verboseCall = verboseCall
```
Why do you need find.verboseCall (as opposed to local verboseCall)? Is it used anywhere ouside of find.lua?

borisfom commented:
@soumith: I have merged upstream changes and added some neat extra refactoring (thanks @ngimel for inception) - I believe it's ready to go.

borisfom commented:
@soumith: I have addressed code review requests (and more). 
Also disabled true fp16 for now - let's roll it first as refactoring and FindEx. Will work a bit more on true fp16 later.

soumith commented:
cool, i'm just testing this out, and will merge now.

soumith commented:
Thanks for the refactoring Boris!
